### PR TITLE
Improved detection when using interfaces in Realm classes

### LINKED ISSUE #2808: Generated RealmProxy class try to access wrong RealmProxy class for internal fields when using an interface hierarchy
#### Goal

I'm trying to take advantage of the new interface `RealmModel` and add some abstraction to my models. My goal is to define an interface for my Realm classes to hide details of the implementation.

#### Steps & Code to Reproduce

Given the following structure (in Kotlin, but should be the same in Java):

```kotlin
interface Address : RealmModel
open class RealmAddress() : Address {
    [...]
}

interface User : RealmModel {
    val addresses: MutableList<Address>
}
open class RealmUser() : User {
    override var addresses: RealmList<Address> = RealmList()
}
```

The generated `RealmUserRealmProxy` try to access `RealmAddressRealmProxy` and its table name by relying on a convention on the type of the field `addresses`, but in this case it fails since it tries to look up `AddressRealmProxy` and `class_Address` instead of `RealmAddressRealmProxy` and `class_RealmAddress` since the type of `addresses` is `RealmList<Address>` instead of `RealmList<RealmAddress>`.

I think the code I'm writing should be correct, if not then probably I misunderstand the purpose of using the `RealmModel` interface instead of extending `RealmObject`...

#### Version of Realm and tooling
Realm version(s): 0.90.1

Android Studio version: N/A

Which Android version and device: N/A

### PR DESCRIPTION
**Pull request by cmelchior**
Fixes #2808 

This PR adds additional checking to the annotation processor to prevent the error reported int #2808.
#2808 would not compile anyway, but this way we can report the real error to the user.

@realm/java

### CODE REVIEW

**Comment by beeender** [comment]
:+1:

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java [hunk 1]
```diff
@@ -156,15 +161,46 @@ private boolean checkForFinalFields() {
...
         }
+
         return true;
     }
 
+    private boolean checkReferenceTypes() {
+        for (VariableElement field : fields) {
+            if (Utils.isRealmModel(field)) {
+                // Check that the referenced type is a concrete class and not an interface
+                TypeElement typeElement = elements.getTypeElement(field.asType().toString());
```
Comment (kneth) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: Common code could be put in a small aux. method (making sure that error message is always the same).

**Comment by kneth** [comment]
:+1: with a minor comment.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java (hunk 1 above)
Comment (cmelchior) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: Not sure I understand this? This particular statement is not used elsewhere and the error message is different: One talks about Lists, the other about Objects.

Comment (kneth) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: Sorry, my mistake.

**Comment by kneth** [comment]
Correcting my mistake: :+1:

**Comment by bmunkholm** [comment]
@emanuelez Can you put an extra pair of eyes on this?

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java [hunk 2]
```diff
@@ -156,15 +161,46 @@ private boolean checkForFinalFields() {
...
                     return false;
                 }
+
+                // Check that the referenced type is a concrete class and not an interface
+                TypeMirror fieldType = field.asType();
+                List<? extends TypeMirror> typeArguments = ((DeclaredType) fieldType).getTypeArguments();
+                String genericCanonicalType = typeArguments.get(0).toString();
+                TypeElement typeElement = elements.getTypeElement(genericCanonicalType);
+                if (typeElement.getSuperclass().getKind() == TypeKind.NONE) {
+                    Utils.error("Only concrete Realm classes are allowed in Lists. Neither " +
```
Comment (nhachicha) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: minor: 
_are allowed in RealmCollection_?

File: realm/realm-annotations-processor/src/test/resources/some/test/InterfaceObjectReference.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+import io.realm.RealmList;
+import io.realm.RealmModel;
+import io.realm.annotations.RealmClass;
+
+@RealmClass
+public class InterfaceObjectReference implements CustomInterface {
+    String name;
+    CustomInterface obj; // Polymorphism, not yet supported.
+}
```
Comment (nhachicha) [inline realm/realm-annotations-processor/src/test/resources/some/test/InterfaceObjectReference.java]: new line for checkstyle

**Comment by nhachicha** [comment]
👍 minor comments

File: realm/realm-annotations-processor/src/test/resources/some/test/InterfaceObjectReference.java (hunk 3 above)
Comment (cmelchior) [inline realm/realm-annotations-processor/src/test/resources/some/test/InterfaceObjectReference.java]: Done

File: CHANGELOG.md [hunk 4]
```diff
@@ -21,6 +22,7 @@
 ### Credits
 
 * Paweł Surówka (@thesurix) for adding the `RealmConfiguration.Builder.assetFile()`.
+>>>>>>> master
```
Comment (kneth) [inline CHANGELOG.md]: Remove

**Comment by emanuelez** [comment]
👍

**Pull request merged**